### DISCUSSION TITLE
Turning a field into a subdocument using TypeHandler<>

### DISCUSSION THREAD

**Discussion by bradsgreen** [jorgen/json_struct#41] [Show and tell]
I puzzled for a bit how to turn a complex object of mine, Price, which contains both a value and a currency type.  It isn't an open struct so using the JS_OBJ() macro isn't possible.  I discovered this solution which might be worth adding to an example for future consumers like me:

(simplified from my actual code)

```
template<>
struct TypeHandler<Price>
{
   struct Px {
      double value;
      std::string isoCurrency;
      JS_OBJ(value, isoCurrency);
   }

   static inline Error to(Price &to_type, ParseContext &context) {
      Px val;
      Error err = TypeHandler<Px>::to(val, context);
      to_type = Price {dval, isoCurrency};
      return err;
   }

   static inline void from(const T &from_type, Token &token, Serialized &serialize) {
      Px px{from_type.doubleVal(), from_type.isoCurrency()};
      return TypeHandler<Px>::from(px, token, serializer);
   }
}
```

**Comment by jorgen** [comment]
Ah! Nice. Yeah, so you keep the declarative serialisation and deserialisation, but with support for immutable objects.

Maybe there is a way to serialize and deserialize into a std::tuple and then pass all the members of the tuple to a constructor based on the order the metadata is defined in... Ie. we could have something like a JS_OBJ_IMMUTABLE or something, and it would use a tuple typehandler based on the metadata registered, and then pass it to a constructor like you do here...

Anyway, nice example!
